software-challenge/navigation.py

```diff
@@ -0,0 +1,139 @@
+"""
+You will be implementing a hypothetical navigation class,
+which keeps track of the boring machine's current position, its desired position,
+and makes any adjustments necessary to keep the machine on track.You will make the
+assumption that two other interfaces are written to make your life easier, the GPS
+class will interface with our onboard GPS sensor to get thecurrent position, and
+the Steering class will control the actuation necessary to move the borer left or
+right (using left, right, and distance to as opposed to actual coordinates
+for simplicity)
+"""
+
+""" VISUALIZATION OF COORDINATE PLANE FOR STEERING
+
+x-Plane
+    behind target --------------- target
+
+y-Plane
+    left of target ---------------target------------------ right of target
+
+z-Plane
+    above target
+    |
+    |
+    |
+    |
+    |
+    target
+    |
+    |
+    |
+    |
+    |
+    below target
+"""
+
+""" Class to control position based on GPS input """
+class Navigation:
+    def __init__(self, GPS, steering):
+
+        # Instances of the two classes defined below for use to complete the navigation method
+        self.GPS = GPS
+        self.steering = steering
+
+        # These will be tuples of the form (x, y, z)
+        self.current_position = None
+        self.desired_position = None
+    
+    """ Sets the desired position that the borer will attempt to move to
+    
+    Note: assume the user will pass in the desired_position parameter when using
+          the interface
+    """
+    def set_desired_position(self, desired_position):
+        pass
+
+    """ Updates the current position """
+    def update_current_position(self):
+        pass
+
+    """ Navigate to the desired position from the current position
+    
+    Based on the current position tuple, compared to the desired position tuple,
+    make decisions on steering, and ensure that the actuations are successful
+
+    Returns: True if actuation requests were successful, False if not
+    Note: It may be good to notify the user if something unexpected happens!
+    """
```

Comment:
These various docstrings should go underneath the function prototype :) I'll fix this up